### ISSUE TITLE
PolicyState

### ISSUE THREAD

**Issue by zsunberg** [sisl/POMDPs.jl#34]
## A policy should map *policy states* to actions

Up until this time, we have conceptualized a policy to be a mapping from beliefs to actions, and we have given responsibility for defining the belief structure and update behavior to the problem writer. However, a more general formulation that is perhaps more appropriate in this context is that a policy maps a policy state (that may be a belief) to an action. This reflects the fact that many advanced POMDP solvers have a closely associated belief representation, so they maintain and update their own state or belief. Thus, belief maintenance should be under control of the solver writer rather than the problem writer (though the problem writer will provide a structure for representing a distribution over states that can be used to represent the initial belief, and may provide exact belief updates for the policy to use if they are available).

The policy state will take different forms for different solvers. For example, the policy state for an MCVI policy is the node in the policy graph, the policy state for a POMCP policy is the action-observation tree and associated particle filter, and the policy state for a policy that feeds a baby when it is crying is simply the previous observation.

In order to implement this new paradigm, we are considering adding three things to the interface (there will probably be other minor changes associated with this as well).

First, we will introduce a new abstract type
```julia
abstract PolicyState
```

Second, we will introduce a function to create an initial policy state at the beginning of the simulation
```julia
function initial_policy_state(policy::Policy, belief::Belief)
```
(also the standard `create_policy_state(p::Policy)` will be declared)

Finally, we will replace the `belief` function in the simulation loop with a function that updates the policy state (which, again, could be a belief representation)
```julia
function update(policy::Policy, ps_old::PolicyState, a::Action, o::Observation, ps_new=create_policy_state(policy))
```

A simulation loop will look like this:

```julia
function simulate(sim::MySimulator, pomdp::POMDP, policy::Policy)

    # initialize belief, state, stopping criteria, allocate stuff etc
    ...

    # create the policy state from the initial belief
    ps = initial_policy_state(policy, belief)
    ps2 = create_policy_state(policy)

    while ... # stopping criteria

        a = action(policy, ps, a)
        r += reward(pomdp, s, a)

        # sample the next state and observation
        trans_dist = transition(pomdp, s, a, trans_dist)
        rand!(sim.rng, s2, trans_dist)
        obs_dist = observation(pomdp, s2, a, obs_dist)
        rand!(sim.rng, o, obs_dist)

        ps2 = update(policy, ps, a, o, ps2)

        # switch things around so we don't have to allocate more
        tmp=s; s=s2; s2=tmp;
        tmpps=ps, ps=ps2, ps2=tmpps;
    end
end
```

## Issues

There are a few issues that we need to nail down before adding this:

### Type Heirarchy

The proper location of `PolicyState` in the type heirarchy is not completely clear. Originally we thought that `Belief` and `PolicyState` should simply be aliased

```julia
typealias PolicyState Belief
```

However, this is a little strange because a `PolicyState` may not be sampleable even though it will then be a subclass of `AbstractDistribution`.

`Belief` should perhaps be a subtype of `PolicyState` because it is a policy state that can be sampled. However, since julia does not allow multiple inheritance, `Belief` could no longer be an `AbstractDistribution`.

Another option is to declare that, `PolicyState` is not connected by inheritance to any other classes, and if a policy state is based on the belief, it should include the belief as a member, i.e.
```
type MyPolicyState <: PolicyState
    b::Belief
end
```

Another wrinkle to think about is that, for an MDP, the policy state could be the MDP state itself, but I think we should just handle this separately with an extra `action(..., s::State,...)` function like we have now.

### Arguments

We need to decide for sure what the arguments for each function should be. In particular, we need to answer whether the following arguments should be in the following functions:

- `pomdp` in `action`? (I think no because the policy should have knowledge of the pomdp from the call to `solve()`)
- `pomdp` in `create_policy_state`? (I think no for the same reason as above)

### Problem Specific Belief Updates

Should POMDPs.jl still contain a standard belief update interface (I think it should), or should we break it out into a different package that can focus more on it? Should it look any different? Does this have any bearing on (#32)?

Does anyone have any thoughts on this?

**Comment by zsunberg** [comment]
While biking back I thought of two things:

1) We could call `update` `observe` or some other more descriptive word
2) Actually I think we should take the exact belief update interface out of POMDPs.jl and put it in another package. Then we can make it as nice and fancy as we want without worrying about cluttering up POMDPs.

**Comment by zsunberg** [comment]
What if we keep `Belief` and `PolicyState` separate in the type hierarchy, but then change the `::PolicyState` argument in both `action` and `update` to be a `::Union(Belief, PolicyState)` argument?

**Comment by etotheipluspi** [comment]
+1 for making a union of Belief and PolicyState. Really good idea!

Also, very nice write-up!

**Comment by zsunberg** [comment]
We could also overload `update` (or could be called `observe`) to act as the standard belief update interface. If #32 is adopted, there could be two function declarations in POMDPs:

```julia
function update(policy::Policy,
                ps_old::Union(PolicyState, Belief),
                a::Action,
                o::Observation,
                ps_new::create_policy_state(policy))
```
and
```julia
function update(updater::BeliefUpdater,
                belief_old::Belief,
                a::Action,
                o::Observation,
                belief_new::create_belief(updater))           
```

In the simulation loop, this function would *always* (this is important!) *always* *always* be called with the policy as the first argument (so the first method would be dispatched). The second method *might* be called within `update(policy, ...)` if the policy needs to perform belief updates if the policy state is a belief or contains beliefs (a default belief updater could be defined by the problem-writer for the policy to use).

Thus the second method defines the interface that the problem-writer could use to communicate exact belief update behaviour to the solver-writer in a standard way (it replaces the `belief` function we have now).

**Comment by ebalaban** [comment]
I need to give this more thought. I think that anytime new terminology or concepts are introduced in a somewhat established field (such as POMDPs), there is a pretty heavy burden of proof to show that the new thing is valuable and cannot be expressed through existing concepts/terminology. I'd like to take a closer look at the solvers mentioned, for example, and see if beliefs/belief updating cannot be done cleanly without introducing PolicyState. Here are a few quick thoughts in the meantime:

- Personally, I've never considered belief update functions to be the responsibility of problem writers. As I mentioned in an earlier comment, I see belief updaters getting created/provided along with solvers in the near future and in dedicated packages in the long term. I also think that a basic belief updating interface should be in POMDPs.jl, since the problem is one of its inputs, but I suppose I can be persuaded to rip it all out and put it in a separate package.

- Even with the latest comments from @zsunberg, it still does not look to me that introducing PolicyState will help solve the issue described in #33.

- If there is an unease with having objects of Solver subtypes as arguments to 'create_belief' or 'initial_belief', we can dispatch on Belief subtypes. That should help in keeping the separation of church and state (eh... belief updaters and solvers) intact.

- While we are on the subject, I think that the names 'belief', 'update', or 'observe' for the belief updating function are too ambiguous. I don't see anything wrong with sparing a few more characters and calling the function 'update_belief', where it will be instantly clear what it does to anyone familiar with the field. I know that the name was shortened to bring it more in line with 'action', 'solve', and such, but to me clarity outranks conciseness (and, frankly, I think that 'get_action' would go better with 'create_action', for example).

**Comment by zsunberg** [comment]
@ebalaban, thanks for your comments - I think it is really useful to have someone act as the current interface's "defense attourney" so that only polished ideas make it into the interface.

I'll try to address some of these concerns briefly.

First, the concept of a policy state is not entirely new. It is present explicitly, for example in the MCVI paper [1], see Section 3.1 "The finite state controller does not maintain the robot’s belief explicitly, as in (3). It encodes the belief implicitly in the controller state based on the robot’s initial belief b and the sequence of observations received."

Second, this can indeed all be expressed using existing concepts/terminology if the policy state is implemented as a `Belief` object as I have been doing with POMCP up to this point. This is strange though because the most convenient way to represent the policy state may not look much like a belief, so the `PolicyState` type might make it clearer. The only hiccup to this is that we are missing a way to automatically convert from the initial belief, which must be specified in problem-specific terms to the initial policy state. I proposed `specialize_belief` in #33 as a band-aid for this, and now we have `initial_policy_state` to fix it.

To address the first bullet: Yes, as you say belief updaters would, at least initially, (and I suspect nearly always, even in the future) be closely tied to solvers. This is why we were thinking that the policy should be the most important player in the policy state/belief update. if the solver can handle different approximate ways of updating beliefs, that can be included in the policy object.

To address the second bullet, I think `initial_policy_state(..., initial_belief(pomdp))` can be used to get the same functionality as `initial_belief` in #33 and `create_policy_state` can function in the same way as `create_belief` in #33. Both of these functions can certainly return a belief because, for many solvers, the policy state will indeed be a belief.

[1] Bai, H., Hsu, D., Lee, W., & Ngo, V. (2011). Monte Carlo value iteration for continuous-state POMDPs. Algorithmic Foundations of Robotics IX, 175–191. Retrieved from http://link.springer.com/chapter/10.1007/978-3-642-17452-0_11

**Comment by mykelk** [comment]
Very nice writeup and discussion. Getting the right design is hard stuff.

Here is an alternative idea that might address some of Edward's and Zach's concerns. It's a little radical and contrary to what I was originally thinking.
0. There is a "problem designer" (the domain expert), a "solver writer" (who writes solvers for generic problems), and an "experimenter" (who runs experiments).
1. Don't have the problem designer worry about belief.
2. Only the solver writer should worry about the belief representation.
3. The experimenter would tell the solver how to initialize (sometimes this would be some distribution, a set of particles, a Gaussian distribution, etc.)
4. Just use Belief instead of PolicyState. The belief is just the state of the policy.
5. Just call the update function `update`. I think this is the more object-oriented and Julian way to do it. We don't have print_array and print_float and print_integer; we just have print.

Here's the example:
```julia
function simulate(sim::MySimulator, pomdp::POMDP, policy::Policy, initial_state::State, belief::Belief)
    # allocate a, o, trans_dist, obs_dist
    s = deep_copy(initial_state)
    while ... # stopping criteria
        a = action(policy, belief, a)
        r += reward(pomdp, s, a)
        trans_dist = transition(pomdp, s, a, trans_dist)
        rand!(rng(sim), s, trans_dist)
        obs_dist = observation(pomdp, s, a, obs_dist) # check ordering of s and a
        rand!(rng(sim), o, obs_dist)
        update(policy, belief, a, o)
    end
end
```

**Comment by ebalaban** [comment]
> Here is an alternative idea that might address some of Edward's and 
> Zach's concerns. It's a little radical and contrary to what I was 
> originally thinking.
>
While undeniably radical, I think this is closer to what I had in mind :-)
>
> 0. There is a "problem designer" (the domain expert), a "solver 
> writer" (who writes solvers for generic problems), and an 
> "experimenter" (who runs experiments).
> 1. Don't have the problem designer worry about belief.
> 2. Only the solver writer should worry about the belief representation.
> 3. The experimenter would tell the solver how to initialize (sometimes 
> this would be some distribution, a set of particles, a Gaussian 
> distribution, etc.)
>
Ok, but how? The problem would still contain the domain knowledge needed 
to form an initial belief, right?  In what form would it be described 
there? How would it get communicated (by the experimenter) to the 
solver? Are you assuming that the experimenter will need to figure out 
the domain knowledge for the initial belief on his/her own? How does it 
get extracted out of the solver (in its new "solver-native" form) to be 
passed to the belief updater?
>
> 4. Just use Belief instead of PolicyState. The belief is just the 
> state of the policy.
>
I agree. I understand that Zach wants to be precise in how we classify 
things, but I don't think that it's that much of a stretch to call what 
MCVI uses a 'belief', for instance. If I understand it correctly, I 
don't see the term 'policy state' adding enough semantically to justify 
its introduction.

> 5. Just call the update function |update|. I think this is the more 
> object-oriented and Julian way to do it. We don't have print_array and 
> print_float and print_integer; we just have print.
>
That's fine. I guess I am still not completely assimilated into the 
Julian way :-) I think a while ago there was a discussion about using 
the convention of the first function argument being the one that's 
getting modified. I realize that this is not always possible (e.g., when 
we have optional arguments), but whenever is, I'd prefer at least that, 
for the sake of readability.

I still don't quite understand the rationale behind storing the belief 
updater object inside the policy object, however. What am I missing? I 
know that for online solvers policy objects are a catch-all kind of a 
thing already anyway, but still, why not have update(belief::Belief, 
bu::BeliefUpdater, a::Any, o::Any), for example? And if access to the 
belief updater is needed within the solver, why not store a reference to 
the belief updater there? Just trying to keep things as intuitive as 
possible...

**Comment by zsunberg** [comment]
@mykelk, right. If you recall, during out meeting on Monday, I think I said "We don't actually need to add anything - we can just call it Belief". So I think that part is fine :). The main conceptual point is that the "belief" that the policy feeds back on will usually be policy-specific.

As @ebalaban says, though, everyone will have to think about belief. The initial belief will often be most naturally specified in a problem-specific domain, so the problem designer should at least provide a belief structure. Moreover, if there is a convenient exact or problem-specific-approximate belief update, there should be a mechanism for the problem designer to specify it (in POMDPs or another package). POMCP, for example, can either use its own particle filter or exact updates to construct the tree.

As you suggest in (3), we should certainly allow the experimenter to specify details of the initial policy-specific belief if they wish (I now see that this was a big part of #33), but we also need to provide a standard function to automatically create a policy-specific belief from a problem-specific belief. The experimenter should not have to learn the structure policy-specific belief if they don't want to. This was `specialize_belief` in #33 or `initial_policy_state` here.

**Comment by zsunberg** [comment]
See my comment on #32 for how to include `BeliefUpdater` in this.

**Comment by mykelk** [comment]
@zsunberg: what if we do some kind of convert(::Type{MCTSBelief}, b::GridWorldBelief)?

**Comment by zsunberg** [comment]
@mykelk, so the experimenter would call `convert(typeof(create_belief(policy)), initial_belief)` either before or within `simulate`? This is a little clunky, but I guess it would work. The point is for the experimenter to not have to look up and type out the name of the policy specific belief structure and replace it every time he wants to try a new solver.

I suppose that `update` could just automatically convert it after the first iteration. i.e. the `belief_old` argument to `update` would be a problem-specific belief, but it would return a policy-specific belief. I couldn't do that before now because the policy was not an argument to `belief`, but now it would work just fine. The more I think about it, the more I am satisfied with just doing the conversion within `update` even though it's not very explicit.

**Comment by mykelk** [comment]
Hmmm... I'm not married to what I suggested. I think Julia can do automatic conversion. Right?

**Comment by zsunberg** [comment]
closing after our in-person discussion friday - will be addressed in #35
